lib/libzfs/libzfs_pool.c

```diff
@@ -4310,3 +4353,197 @@ zpool_label_disk(libzfs_handle_t *hdl, zpool_handle_t *zhp, char *name)
 
 	return (0);
 }
+
+#if HAVE_LIBDEVMAPPER
+static void libdevmapper_dummy_log(int level, const char *file, int line,
+    int dm_errno_or_class, const char *f, ...) {}
+
+/* Disable libdevmapper error logging */
+static void disable_libdevmapper_errors(void) {
+	dm_log_with_errno_init(libdevmapper_dummy_log);
+}
+/* Enable libdevmapper error logging */
+static void enable_libdevmapper_errors(void) {
+	dm_log_with_errno_init(NULL);
+}
+#endif
+
+/*
+ * Allocate and return the underlying device name for a device mapper device.
+ * If a device mapper device maps to multiple devices, return the first device.
+ *
+ * For example, dm_name = "/dev/dm-0" could return "/dev/sda"
+ *
+ * dm_name should include the "/dev[/mapper]" prefix.
+ *
+ * Returns device name, or NULL on error or no match.  If dm_name is not a DM
+ * device then return NULL.
+ *
+ * NOTE: The returned name string must be *freed*.
+ */
+static char * dm_get_underlying_path(char *dm_name)
+{
+	char *name = NULL;
+#if HAVE_LIBDEVMAPPER
+	char *tmp;
+	struct dm_task *dmt = NULL;
+	struct dm_tree *dt = NULL;
+	struct dm_tree_node *root, *child;
+	void *handle = NULL;
+	struct dm_info info;
+	const struct dm_info *child_info;
+	int size;
+
+	/*
+	 * Disable libdevmapper errors.  It's entirely possible user is not
+	 * running devmapper, or that dm_name is not a devmapper device.
+	 * Thats totally ok, we will just harmlessly and silently return NULL.
+	 */
+	disable_libdevmapper_errors();
+
+	/*
+	 * libdevmapper tutorial
+	 *
+	 * libdevmapper is basically a fancy wrapper for its ioctls.  You
+	 * create a "task", fill in the needed info to the task (fill in the
+	 * ioctl fields), then run the task (call the ioctl).
+	 */
+	/* First we need the major/minor number for our DM device */
+	if (!(dmt = dm_task_create(DM_DEVICE_INFO)))
+		goto end;
+
+
+	/* Lookup the name in libdevmapper */
+	if (!dm_task_set_name(dmt, dm_name)) {
+		enable_libdevmapper_errors();
+		goto end;
+	}
+
+	if (!dm_task_run(dmt))
+		goto end;
+
+	/* Get DM device's major/minor */
+	if (!dm_task_get_info(dmt, &info))
+		goto end;
+
+	/* We have major/minor number.  Lookup the dm device's children */
+	if (!(dt = dm_tree_create()))
+		goto end;
+
+	/* We add the device into the tree and it's children get populated */
+	if (!dm_tree_add_dev(dt, info.major, info.minor))
+		goto end;
+
+	if (!(root = dm_tree_find_node(dt, 0, 0)))
+		goto end;
+
+	if (!(child = dm_tree_next_child(&handle, root, 1)))
+		goto end;
+
+	/* Get child's major/minor numbers */
+	if (!(child_info = dm_tree_node_get_info(child)))
+		goto end;
+
+	/* Got childs major/minor numbers.  Translate it into a device */
+	size = snprintf(NULL, 0, "/dev/block/%d:%d",
+	    child_info->major, child_info->minor);
+	/* +1 for '\0' terminator */
+
+	if (!(tmp = malloc(size)))
+		goto end;
+
+	sprintf(tmp, "/dev/block/%d:%d", child_info->major, child_info->minor);
```

Comment:
So I took a look at illumos and while asprintf() is a GCC extension it is available.  From the man page.

```
       See standards(5) for  the  standards  conformance  of  printf(),  fprintf(),  sprintf(),  and
       snprintf().  The  asprintf() function is modeled on the one that appears in the FreeBSD, Net‐
       BSD, and GNU C libraries.
```

Let's not make our lives needlessly difficult.  Just use it.